In [3]:
import numpy as np
np.set_printoptions(formatter={'float': lambda x: "{:.2f}".format(x)}) # configura a impressão dos floats

# Métodos Numéricos

## Resolução de Sistemas Lineares
## O método de Gauss com Pivotação

### Patric Lacouth

## O problema

tempo, t | velocidade, v
-----|------
(s)|(m/s)
5|106.8
8|177.2
12|279.2

$$v(t) = at^2 + bt + c$$

## O sistema

$$\left\{\begin{matrix}
25a & + & 5b & + & c & = & 106.8 \\ 
64a & + & 8b & + & c & = & 177.2 \\ 
144a & + & 12b & + & c & = & 279.2 
\end{matrix}\right.$$

## Sistemas Equivalentes

$$\left\{\begin{matrix}
25a & + & 5b & + & c & = & 106.8 \\ 
64a & + & 8b & + & c & = & 177.2 \\ 
144a & + & 12b & + & c & = & 279.2 
\end{matrix}\right.$$

$$\left\{\begin{matrix}
25a & + & 5b & + & c & = & 106.8 \\ 
 &  & 4.80b & - & 1.56c & = & -96.21 \\ 
 &  &  &  & 0.70c & = & 0.76 
\end{matrix}\right.$$

# Método de Gauss com Pivotação Parcial

* Método **direto**

* Método matricial

# As etapas

* Eliminação progressiva

 * Substituição retroativa

<img src = "imagens/exemplo_eliminacao.png" width = 50% style = "display : block; margin-right : auto; margin-left : auto;" />

# Operações elementares

1. multiplicação de um linha por uma constante não nula.

2. somar linhas.

3. permutação de duas linhas.

# Representação Matricial

$$\left\{\begin{matrix}
25x_1 & + & 5x_2 & + & x_3 & = & 106.8 \\ 
64x_1 & + & 8x_2 & + & x_3 & = & 177.2 \\ 
144x_1 & + & 12x_2 & + & x_3 & = & 279.2 
\end{matrix}\right.$$

$$\left[ \begin{matrix}
25 & 5 & 1 \\
64 & 8 & 1\\
144 & 12 & 1
\end{matrix} \right]
\left[ \begin{matrix}
x_{1}\\
x_{2}\\
x_{3}
\end{matrix} \right]
=
\left[ \begin{matrix}
106.8\\
177.2\\
279.2
\end{matrix} \right]
$$

$$[A][x] = [b]$$

# Matriz Estendida

como vamos fazer operações com as matrizes $[A]$ e $[b]$ no python podemos representar as duas matrizes através de uma matriz estendida que contém todos os valores



$$\left[ \begin{matrix}
25 & 5 & 1 \\
64 & 8 & 1\\
144 & 12 & 1
\end{matrix} \right]
\left[ \begin{matrix}
x_{1}\\
x_{2}\\
x_{3}
\end{matrix} \right]
=
\left[ \begin{matrix}
106.8\\
177.2\\
279.2
\end{matrix} \right]
$$

$$[AB] = \left[ \begin{matrix}
25 & 5 & 1 & 106.8\\
64 & 8 & 1 & 177.2\\
144 & 12 & 1 & 279.2
\end{matrix} \right]$$

# O algoritmo da eliminação progressiva

1. Determinar o **pivô**, escolhendo o de maior valor absoluto na coluna

$$[AB] = \left[ \begin{matrix}
25 & 5 & 1 & 106.8\\
64 & 8 & 1 & 177.2\\
144 & 12 & 1 & 279.2
\end{matrix} \right]$$

2. Criar o multiplicador, **$m_x$** , para escalonar a linha, $m_1 = \frac{64}{144}$

$$[AB] = \left[ \begin{matrix}
144 & 12 & 1 & 279.2 \\
64 & 8 & 1 & 177.2\\
25 & 5 & 1 & 106.8
\end{matrix} \right]$$

3. Subtrair a linha do elemento a ser zerado pela linha do pivô escalonada pelo multiplicador.

4. Repetir o processo para todos os valores a serem eliminados

# Implementando Gauss em Python

## Importando as bibliotecas

In [4]:
import numpy as np

## Representando o sistema no python

$$[AB] = \left[ \begin{matrix}
25 & 5 & 1 & 106.8\\
64 & 8 & 1 & 177.2\\
144 & 12 & 1 & 279.2
\end{matrix} \right]$$

In [5]:
AB = np.array([[25,5,1,106.8],[64,8,1,177.2],[144,12,1,279.2]], dtype = float)
AB

array([[25.00, 5.00, 1.00, 106.80],
       [64.00, 8.00, 1.00, 177.20],
       [144.00, 12.00, 1.00, 279.20]])

### Selecionando o primeiro pivô

Precisamos "zerar" os elementos abaixo do primeiro elemento da primeira coluna, para isso escolhemos o elemento com maior módulo na coluna 0 para ser o pivô

$$[AB] = \left[ \begin{matrix}
25 & 5 & 1 & 106.8\\
64 & 8 & 1 & 177.2\\
144 & 12 & 1 & 279.2
\end{matrix} \right]$$

In [6]:
AB[[0,2]] = AB[[2,0]]
AB

array([[144.00, 12.00, 1.00, 279.20],
       [64.00, 8.00, 1.00, 177.20],
       [25.00, 5.00, 1.00, 106.80]])

In [7]:
pivo = AB[0,0]
pivo

144.0

Determinamos o primeiro multiplicador **m1** como sendo o elemento que queremos eliminar dividido pelo nosso pivô

In [8]:
AB

array([[144.00, 12.00, 1.00, 279.20],
       [64.00, 8.00, 1.00, 177.20],
       [25.00, 5.00, 1.00, 106.80]])

In [9]:
m1 = AB[1,0] / pivo
m1

0.4444444444444444

Para zerar o elemento abaixo do pivô iremos subtrair a sua linha pela linha do pivô multiplicada por **m1**

In [10]:
AB

array([[144.00, 12.00, 1.00, 279.20],
       [64.00, 8.00, 1.00, 177.20],
       [25.00, 5.00, 1.00, 106.80]])

In [11]:
AB[1] = AB[1] - AB[0] * m1

In [12]:
AB

array([[144.00, 12.00, 1.00, 279.20],
       [0.00, 2.67, 0.56, 53.11],
       [25.00, 5.00, 1.00, 106.80]])

Para zerar o próximo elemento da coluna, precisamos de um novo multiplicador, **m2**, que será o próprio elemento dividido pelo pivô

In [13]:
AB

array([[144.00, 12.00, 1.00, 279.20],
       [0.00, 2.67, 0.56, 53.11],
       [25.00, 5.00, 1.00, 106.80]])

In [14]:
m2 = AB[2,0] / pivo
m2

0.1736111111111111

Repetimos o processo de subtrair a linha na qual desejamos zerar o elemento pela linha do pivô multiplicada por **m2**

In [15]:
AB

array([[144.00, 12.00, 1.00, 279.20],
       [0.00, 2.67, 0.56, 53.11],
       [25.00, 5.00, 1.00, 106.80]])

In [16]:
AB[2] = AB[2] - AB[0] * m2
AB

array([[144.00, 12.00, 1.00, 279.20],
       [0.00, 2.67, 0.56, 53.11],
       [0.00, 2.92, 0.83, 58.33]])

## Selecionando o segundo pivô

Depois de "eliminar" os elementos da primeira coluna temos que continuar a eliminação na segunda coluna. O pivô agora passa a ser o elemento com maior valor absoluto na segunda coluna 

## Selecionando o segundo pivô

In [17]:
AB

array([[144.00, 12.00, 1.00, 279.20],
       [0.00, 2.67, 0.56, 53.11],
       [0.00, 2.92, 0.83, 58.33]])

In [18]:
AB[[1,2]] = AB[[2,1]]
AB

array([[144.00, 12.00, 1.00, 279.20],
       [0.00, 2.92, 0.83, 58.33],
       [0.00, 2.67, 0.56, 53.11]])

In [19]:
pivo = AB[1,1]
pivo

2.916666666666667

Nesta coluna temos que realizar apenas a eliminação de um elemento abaixo do pivô, repetimos o mesmo processo anterior, calculamos um multiplicador **m3** a partir da divisão do elemento a ser zerado pelo pivô

In [20]:
AB

array([[144.00, 12.00, 1.00, 279.20],
       [0.00, 2.92, 0.83, 58.33],
       [0.00, 2.67, 0.56, 53.11]])

In [21]:
m3 = AB[2,1] / pivo
m3

0.9142857142857143

Com o multiplicador, **$m_3$** realizamos a operação na linha a ser modificada

In [22]:
AB

array([[144.00, 12.00, 1.00, 279.20],
       [0.00, 2.92, 0.83, 58.33],
       [0.00, 2.67, 0.56, 53.11]])

In [23]:
AB[2] = AB[2] - AB[1] * m3
AB

array([[144.00, 12.00, 1.00, 279.20],
       [0.00, 2.92, 0.83, 58.33],
       [0.00, 0.00, -0.20, -0.22]])

## Substituição: encontrando os valores de x

Para encontrar $x_3$ precisamos apenas resolver a equação representada pela última linha do sistema

$$\left\{\begin{matrix}
144x_1 & + & 12x_2 & + & x_3 & = & 279.2 \\ 
 &  & 2.92x_2 & + & 0.83x_3 & = & 58.33 \\ 
 &  &  &  & -0.2x_3 & = & -0.22 
\end{matrix}\right.$$

## Criando um vetor vazio X

In [24]:
x = np.zeros(3)
x

array([0.00, 0.00, 0.00])

### Encontrando o valor de $x_3$
Para encontrar $x_3$ precisamos apenas resolver a equação representada pela última linha do sistema

$$-0.2x_3 = -0.22$$

$$x_3 = \frac{-0.22}{-0.2}$$

In [25]:
AB

array([[144.00, 12.00, 1.00, 279.20],
       [0.00, 2.92, 0.83, 58.33],
       [0.00, 0.00, -0.20, -0.22]])

In [26]:
x[2] = AB[2,3] / AB[2,2]
x

array([0.00, 0.00, 1.09])

### Encontrando o valor de $x_2$
Com o valor de $x_3$ definido podemos encontrar $x_2$

$$2.92x_2 + 0.83x_3 = 58.33$$


$$x_2 = \frac{58.33 - 0.83x_3}{2.92}$$

In [28]:
AB

array([[144.00, 12.00, 1.00, 279.20],
       [0.00, 2.92, 0.83, 58.33],
       [0.00, 0.00, -0.20, -0.22]])

In [29]:
x[1] = (AB[1,3] - AB[1,2] * x[2]) / AB[1,1]
x

array([0.00, 19.69, 1.09])

### Encontrando o valor de $x_1$
Com os valores de $x_3$ e $x_2$ definidos podemos encontrar $x_1$

$$144x_1 + 12x_2 + x_3 = 279.20$$

$$x_1 = \frac{279.2 - 1x_3 - 12x_2}{144}$$

In [31]:
AB

array([[144.00, 12.00, 1.00, 279.20],
       [0.00, 2.92, 0.83, 58.33],
       [0.00, 0.00, -0.20, -0.22]])

In [32]:
x[0] = (AB[0,3] - AB[0,2] * x[2] - AB[0,1] * x[1]) / AB[0,0]
x

array([0.29, 19.69, 1.09])

### Verificando a resposta
$$\left[ \begin{matrix}
25 & 5 & 1 \\
64 & 8 & 1\\
144 & 12 & 1
\end{matrix} \right]
\left[ \begin{matrix}
x_{1}\\
x_{2}\\
x_{3}
\end{matrix} \right]
=
\left[ \begin{matrix}
106.8\\
177.2\\
279.2
\end{matrix} \right]
$$

$$[A][x] = [b]$$

In [33]:
A = np.array([[25,5,1],[64,8,1],[144,12,1]],dtype = float)
A

array([[25.00, 5.00, 1.00],
       [64.00, 8.00, 1.00],
       [144.00, 12.00, 1.00]])

In [34]:
x

array([0.29, 19.69, 1.09])

In [35]:
b = np.dot(A,x)
b

array([106.80, 177.20, 279.20])

# Now, go code!